aydin/nn/models/torch/torch_jinet.py

```diff
@@ -0,0 +1,291 @@
+        lr=learning_rate,
+        start_noise_level=training_noise,
+        weight_decay=l2_weight_regularization,
+    )
+
+    scheduler = ReduceLROnPlateau(
+        optimizer,
+        'min',
+        factor=reduce_lr_factor,
+        verbose=True,
+        patience=reduce_lr_patience,
+    )
+
+    def loss_function(u, v):
+        return torch.abs(u - v)
+
+    for epoch in range(nb_epochs):
+        train_loss_value = 0
+        validation_loss_value = 0
+        iteration = 0
+        for i, (input_image, target_image) in enumerate(
+            zip([input_images], [target_images])
+        ):
+            optimizer.zero_grad()
+
+            model.train()
+
+            translated_image = model(input_image)
+
+            translation_loss = loss_function(translated_image, target_image)
+
+            translation_loss_value = translation_loss.mean()
+
+            translation_loss_value.backward()
+
+            optimizer.step()
+
+            train_loss_value += translation_loss_value.item()
+            iteration += 1
+
+            # Validation:
+            with torch.no_grad():
+                model.eval()
+
+                translated_image = model(input_image)
+
+                translation_loss = loss_function(translated_image, target_image)
+
+                translation_loss_value = translation_loss.mean().cpu().item()
+
+                validation_loss_value += translation_loss_value
+                iteration += 1
+
+        train_loss_value /= iteration
+        lprint(f"Training loss value: {train_loss_value}")
+
+        validation_loss_value /= iteration
+        lprint(f"Validation loss value: {validation_loss_value}")
+
+        writer.add_scalar("Loss/train", train_loss_value, epoch)
+        writer.add_scalar("Loss/valid", validation_loss_value, epoch)
+
+        scheduler.step(validation_loss_value)
+
+        if validation_loss_value < best_val_loss_value:
+            lprint("## New best val loss!")
+            if validation_loss_value < best_val_loss_value - patience_epsilon:
+                lprint("## Good enough to reset patience!")
+                patience_counter = 0
+
+            best_val_loss_value = validation_loss_value
+
+            best_model_state_dict = OrderedDict(
+                {k: v.to('cpu') for k, v in model.state_dict().items()}
+            )
+        else:
+            if epoch % max(1, reload_best_model_period) == 0 and best_model_state_dict:
+                lprint("Reloading best models to date!")
+                model.load_state_dict(best_model_state_dict)
+
+            if patience_counter > patience:
+                lprint("Early stopping!")
+                break
+
+            lprint(
+                f"No improvement of validation losses, patience = {patience_counter}/{patience}"
+            )
+            patience_counter += 1
+
+        lprint(f"## Best val loss: {best_val_loss_value}")
+
+    writer.flush()
+    writer.close()
+
+
+# def n2s_jinet_train_loop():
```

Comment:
Is this implementation also going to be included in the current PR?

Reply:
It can  be a separate one as it needs more testing